### DISCUSSION TITLE
Change parameters type for custom tuning

### DISCUSSION THREAD

**Discussion by ppyarpe** [microsoft/FLAML#500] [Q&A]
I am currently using FLAML  with custom tunning for a DPBoost (https://github.com/QinbinLi/DPBoost) implementation. 

DPBoost overrides lightgbm and hence to use it ones uses the same functions as LightGBM.

When I run lgb.train(params,data) I get the following error:
--------------------------------------------------------------------------------------------------
  File "<ipython-input-38-c27e087accde>", line 1, in <module>
    model=lgb.train(params,data)

  File "C:\Users\Rita\AppData\Roaming\Python\Python37\site-packages\lightgbm\engine.py", line 156, in train
    if num_boost_round <= 0:

TypeError: '<=' not supported between instances of 'Integer' and 'int'
--------------------------------------------------------------------------------------------------

Params:
{'num_leaves': <flaml.tune.sample.Integer at 0x185a255ef28>,
 'max_depth': <flaml.tune.sample.Integer at 0x185a255eda0>,
 'learning_rate': <flaml.tune.sample.Float at 0x185a255e630>,
 'num_iterations': <flaml.tune.sample.Integer at 0x185a255e780>,
 'my_n_trees': <flaml.tune.sample.Integer at 0x185a255edd8>,
 'lambda_l2': <flaml.tune.sample.Float at 0x185a253a048>,
 'bagging_freq': <flaml.tune.sample.Integer at 0x185a3c11c50>,
 'bagging_fraction': <flaml.tune.sample.Float at 0x185a253a160>,
 'max_bin': <flaml.tune.sample.Integer at 0x185a253a0b8>,
 'inner_boost_round': <flaml.tune.sample.Integer at 0x185a253a1d0>,
 'total_budget': 1,
 'verbose': -2,
 'boosting_type': 'gbdt',
 'objective': 'regression',
 'metric': 'binary_error',
 'boost_method': 'DPBoost_2level',
 'high_level_boost_round': 1,
 'balance_partition': 0,
 'geo_clip': 1,
 'boost_from_average': False}

where the flaml.tune objects come from the configurations and search space.

And data is a lightgbm Dataset.

Is there any way to change the flaml.tune objects into different types in order to feed into the lgb.train function from DPBoost ?

When I try int(params['num_leaves']) ,for example, I get the following error:
---------------------------------------------------------------------------------------------------
Traceback (most recent call last):

  File "<ipython-input-41-c8054f39f7d9>", line 1, in <module>
    int(params['num_leaves'])

TypeError: int() argument must be a string, a bytes-like object or a number, not 'Integer'
-----------------------------------------------------------------------------------------------------------

**Comment by sonichi** [comment]
The `params` in your code seems to be a search space instead of a specific config. Could you share the code that causes the error?

**Reply by ppyarpe** [reply]
import pandas as pd
import numpy as np
from flaml import tune
from flaml.model import LGBMEstimator
import lightgbm as lgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score, mean_squared_error,accuracy_score
from sklearn.preprocessing import MinMaxScaler


## Define dataset
adult = pd.read_csv('adult.csv')

# Read in dataset

variables_adu=(adult.columns)

#transform categorical variables to be compatible with LightGBM
X = adult.drop(columns=['income'])
#X=X.drop(columns=[variables_dia[0]])
y=adult.income.to_frame()
for c in X.columns:
    col_type = X[c].dtype
    if col_type == 'object' or col_type.name == 'category':
        X[c] = X[c].astype('category')
        
y.loc[y['income']=='>50K','income']=1
y.loc[y['income']=='<=50K','income']=-1
y=y.values.ravel()

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=314)

scaler1 = MinMaxScaler(feature_range=(-1,1))
scaler1.fit(y_test.reshape(-1,1))
y_test_sc = scaler1.transform(y_test.reshape(-1,1))


x=X_train
y=y_train

scaler = MinMaxScaler(feature_range=(-1,1))
scaler.fit(y.reshape(-1,1))
y_new = scaler.transform(y.reshape(-1,1))



## Define training function, combining fixed params with params from search space
    
def train_lgbm(config,x=x,y_new=y_new,X_test=X_test,y_test_sc=y_test_sc) :   
    params_fixed = {
        'boosting_type': 'gbdt',
        'objective': 'regression',
        'metric': 'binary_error',
        'total_budget': total_budget,
        'boost_method': 'DPBoost_2level',
        'high_level_boost_round': 1,
        'balance_partition': 0,
        'geo_clip': 1,
        'verbose': -2,
        'boost_from_average': False,
        
    }
    # convert config dict to lgbm params
    params = LGBMEstimator(**config).params
    params.update(params_fixed)
    # train the model
    data = lgb.Dataset(x,y_new.reshape(-1))
    model=lgb.train(params,data,10)
    y_pred=model.predict(X_test)
    y_pred[y_pred<=0]=-1
    y_pred[y_pred>0]=1
    y_test_sc = y_test_sc.astype(int)
    roc=roc_auc_score(y_test_sc,y_pred)
    #acc=accuracy_score(y_test_sc,y_pred)
    #print('roc_auc', '=',roc_auc_score(y_test_sc,y_pred))
    #print('acc', '=',acc)
    #return roc,acc
    return {"roc": roc}
    #return {"acc": acc}
    
 ## Define function to refit model after parameter tuning and return predicted class   
def refit_best_hp(config,x=x,y_new=y_new,X_test=X_test,y_test_sc=y_test_sc) :   
    params_fixed = {
        'boosting_type': 'gbdt',
        'objective': 'regression',
        'metric': 'binary_error',
        'boost_method': 'DPBoost_2level',
        'high_level_boost_round': 1,
        'balance_partition': 0,
        'geo_clip': 1,
        'verbose': -2,
        'boost_from_average': False,
        
    }
    # convert config dict to lgbm params
    params = LGBMEstimator(**config).params
    params.update(params_fixed)
    # train the model
    data = lgb.Dataset(x,y_new.reshape(-1))
    model=lgb.train(params,data,10)
    y_pred=model.predict(X_test)
    y_pred[y_pred<=0]=-1
    y_pred[y_pred>0]=1
    y_test_sc = y_test_sc.astype(int)
    return y_pred
 

## Define function for parameter tuning
def test_tune_lgbm_csv(total_budget):

    config_search_space={
        'num_leaves': tune.qrandint(1, 100),
        'max_depth': tune.qrandint(1, 15),
        'learning_rate': tune.uniform(0.001, 1),
        'num_iterations': tune.qrandint(1, 30000),
        'my_n_trees': tune.qrandint(1, 30000),
        'lambda_l2': tune.uniform(0.001, 1024),
        'bagging_freq': tune.qrandint(0, 10),
        'bagging_fraction':tune.uniform(0.001, 1),
        'max_bin': tune.qrandint(2, 1000),
        'inner_boost_round': tune.qrandint(1, 100),
        'total_budget': total_budget,
            }
    
    low_cost_partial_config={
            'num_iterations':4,
            'num_leaves':4,}
            
    
    points_to_evaluate=[{
        'num_leaves': 31,
        'max_depth': 6,
        'learning_rate': 0.1,
        'num_iterations': 50,
        'my_n_trees': 50,
        'lambda_l2': 0.1,
        'bagging_freq': 1,
        'bagging_fraction':0.5,
        'max_bin': 255,
        'inner_boost_round': 50,
        'total_budget': total_budget,
    }]
    # run the tuning, minimizing mse, with total time budget 3 seconds
    analysis = tune.run(
        train_lgbm,
        metric="roc",
        mode="max",
        config=config_search_space,
        low_cost_partial_config=low_cost_partial_config,
        points_to_evaluate=points_to_evaluate,
        time_budget_s=3,
        num_samples=-1,
    )
    return analysis
    
## Run parameter tuning for different budget lists

total_budgets_list = [1]


perf=pd.DataFrame(columns=['Budget','Accuracy','ROC'])

for total_budget in total_budgets_list:
    analysis=test_tune_lgbm_csv(total_budget) #find best hp
    y_pred=refit_best_hp(config=analysis.best_config) #refits model for best hp and gets y_pred
    roc=roc_auc_score(y_test_sc,y_pred)
    acc=accuracy_score(y_test_sc,y_pred)
                       
    perf = perf.append({'Budget': total_budget,'Accuracy':acc,'ROC':roc}, ignore_index=True)
    
    output_path='y_pred_budget_'+str(total_budget)+'.csv'
    y_pred=pd.DataFrame(y_pred)  
    y_pred.to_csv(output_path)
   
    
    
perf.to_csv('metrics_experiments_epsillon.csv',index=None)
print(perf)
X_train=pd.DataFrame(X_train)  
X_test=pd.DataFrame(X_test)
y_train=pd.DataFrame(y_new) 
y_test=pd.DataFrame(y_test_sc)  


X_train.to_csv('X_train.csv')
X_test.to_csv('X_test.csv')
y_train.to_csv('y_train.csv')
y_test.to_csv('y_test.csv')

**Reply by qingyun-wu** [reply]
Hi @[ppyarpe](https://github.com/ppyarpe), with the provided code above, I do not get any error. Just some warnings from LightGBM such as`[LightGBM] [Warning] Unknown parameter: my_n_trees`.  I am testing with flamlv0.10.0 and python 3.8. Do you still have the error you described? 

Thank you!

**Reply by ppyarpe** [reply]
Hi @qingyun-wu, yes I am still getting the same error when using DPBoost. However installing flaml after DPBoost (which needs to be done using Github specifications) will rewrite DPBoost and hence revert to LightGBM in which case I don't get the error.

To double check if DPBoost is working one should have different metric results for different total_budget

**Reply by qingyun-wu** [reply]
Hi @ppyarpe , I am trying install DPBoost following the instruction from the [link](https://github.com/QinbinLi/DPBoost#script) you provided, but I get the following error when installing it. 
```
  File "setup.py", line 337, in <module>
    'Topic :: Scientific/Engineering :: Artificial Intelligence'])
  File "/usr/lib/python3.7/site-packages/setuptools/__init__.py", line 165, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib64/python3.7/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib64/python3.7/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/usr/lib64/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "setup.py", line 261, in run
    openmp_include_dir=self.openmp_include_dir, openmp_library=self.openmp_library)
  File "setup.py", line 200, in compile_cpp
    silent_call(cmake_cmd, raise_error=True, error_msg='Please install CMake and all required dependencies first')
  File "setup.py", line 86, in silent_call
    raise Exception("\n".join((error_msg, LOG_NOTICE)))
Exception: Please install CMake and all required dependencies first
The full version of error log was saved into /root/LightGBM_compilation.log
```
I tried re-installing cmake and gcc, but still getting the same error.  Any idea how should I resolve this error during DPBoost installation? 

And about `To double check if DPBoost is working one should have different metric results for different total_budget`: I am not sure how should I interpret it.  Do you mean after successfully installing DPBoost, I should revise the code such that there are different metric results for different total_budget to test if DPBoost is working? Can you revise your code to reflect it such that I can directly run your code? 

Thank you!

**Reply by qingyun-wu** [reply]
@ppyarpe, like @sonichi said, the `params` in your code seem to be a flaml search space instead of a specific config. Since I cannot test your code directly now, you might want to double-check whether the input `params` in `model=lgb.train(params, data)`, and the `config` in `LGBMEstimator(**config).params` are properly obtained. It will also be helpful if you can share your whole error stack trace. 
Thank you!

**Reply by ppyarpe** [reply]
@qingyun-wu, regarding the installation of DPBoost I am not sure when I install it in Windows I have downloaded Visual Studio 16 and that has worked and when installing it on a mac I have installed cmake, gcc version 7 and libomp.

Thank you I will double check the params input ! :)
